microsoft/testsuites/kvm/kvm_unit_tests_tool.py

```diff
@@ -47,15 +72,61 @@ def run_tests(self, failure_logs_path: Path) -> List[str]:
             no_info_log=False,  # print out result of each test
         )
 
-        lines = result.stdout.split("\n")
-        failures: List[str] = []
+        results = self._parse_results(exec_result.stdout)
+
+        failed_tests = []
+        for result in results:
+            if result.status == TestStatus.FAILED:
+                failed_tests.append(result.name)
+            community_message = create_test_result_message(
+                CommunityTestMessage,
+                test_result.id_,
+                environment,
+                result.name,
+                result.status,
+            )
+
+            notifier.notify(community_message)
+
+        self._save_logs(failed_tests, failure_logs_path)
+        assert_that(failed_tests, f"Unexpected failures: {failed_tests}").is_empty()
+
+    def _parse_results(self, output: str) -> List[KvmUnitTestResult]:
+        lines = output.split("\n")
+        results: List[KvmUnitTestResult] = []
+
+        # Each line is printed in this format:
+        #
+        # PASS kvm (<some additional info...>)
+        # |    |
+        # |    +-> test name
+        # +-> test status (can also be FAIL or SKIP)
+        #
+        # For now, we don't do anything with the additional info in the
+        # parantheses.
         for line in lines:
+            result = KvmUnitTestResult()
             line = "".join(filter(lambda c: c in string.printable, line))
-            if "FAIL" in line:
-                test_name = line.split(" ")[1]
-                failures.append(test_name)
-        self._save_logs(failures, failure_logs_path)
-        return failures
+            parts = line.split(" ")
+
+            result.name = parts[1]
+
+            status = parts[0]
+            if status == "PASS":
+                result.status = TestStatus.PASSED
+            elif status == "FAIL":
+                if result.name in self.EXPECTED_FAILURES:
+                    result.status = TestStatus.ATTEMPTED
+                else:
+                    result.status = TestStatus.FAILED
+            elif status == "SKIP":
+                result.status = TestStatus.SKIPPED
+            else:
+                continue
```

Comment:
Isn't this issue handled in line 113-115 which ignores any lines not following pattern: "<name> <status>"? We need to ensure that any changes in the script can be detected and adding it to logs might not be enough since they can easily get ignored. However, sending an error here could lead to later test result community messages being skipped. How about we throw an error right after all the community messages are sent?

Also, does each of these result line follow a specific regex pattern? If so, it would be more robust to parse from regex. You can look at the example here: https://github.com/microsoft/lisa/blob/43398b083687a5b1743e2814b03f5d8457157d37/microsoft/testsuites/ltp/ltp.py#L329-L345

Reply:
Sure, I can use a regex pattern but that still leaves the question of what to do with lines that don't match the pattern. 

> We need to ensure that any changes in the script can be detected

No, we don't need to detect all changes in the script. Some changes are harmless for us. If the script starts printing some added info, we can simply ignore that. For example, let's say tomorrow the script prints a summary line at the end. We don't want LISA to throw an error for that because the summary line doesn't match the regex. We can simply ignore that line because we don't need it.

The type of change we _do_ want to detect is if the format of the test status line ever changes. This detection is already done. If the test status line format changes, none of the lines will match the regex (or in the current code all lines will fall to the `else` case) and the results will be empty. There is code above that checks if the `results` list is empty and throws an exception. This exception will alert us.

Reply:
Added regex based parsing.